File: 5 - Coding Interview Workshops/Week12/Source Code/lesson.py
import threading
import queue

# Locks and Mutexes
lock = threading.Lock()

def critical_section():
    lock.acquire()
    try:
        # Access shared resources
        pass
    finally:
        lock.release()

# Semaphores
semaphore = threading.Semaphore(2)

def resource_access():
    semaphore.acquire()
    try:
        # Access shared resources
        pass
    finally:
        semaphore.release()

# Condition Variables
condition = threading.Condition()
data_available = False

def producer():
    global data_available
    with condition:
        data_available = True
        condition.notify()

def consumer():
    with condition:
        while not data_available:
            condition.wait()
        # Consume data

# Scheduling Algorithms
def round_robin(processes, quantum):
    ready_queue = queue.Queue()
    for process in processes:
        ready_queue.put(process)

    while not ready_queue.empty():
        process = ready_queue.get()
        # Execute process for the given quantum
        # If process is not finished, put it back in the queue
        if not process.is_finished():
            ready_queue.put(process)

def priority_scheduling(processes):
    ready_queue = queue.PriorityQueue()
    for process in processes:
        ready_queue.put((process.priority, process))

    while not ready_queue.empty():
        _, process = ready_queue.get()
        # Execute process until completion or preemption

# Interview-Style Questions
class Singleton:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        if not cls._instance:
            with cls._lock:
                if not cls._instance:
                    cls._instance = super().__new__(cls)
        return cls._instance

class ConcurrentQueue:
    def __init__(self):
        self._queue = []
        self._lock = threading.Lock()
        self._condition = threading.Condition(self._lock)

    def put(self, item):
        with self._lock:
            self._queue.append(item)
            self._condition.notify()

    def get(self):
        with self._lock:
            while not self._queue:
                self._condition.wait()
            return self._queue.pop(0)

class ReaderWriterLock:
    def __init__(self):
        self._read_lock = threading.Lock()
        self._write_lock = threading.Lock()
        self._readers = 0

    def acquire_read(self):
        with self._read_lock:
            self._readers += 1
            if self._readers == 1:
                self._write_lock.acquire()

    def release_read(self):
        with self._read_lock:
            self._readers -= 1
            if self._readers == 0:
                self._write_lock.release()

    def acquire_write(self):
        self._write_lock.acquire()

    def release_write(self):
        self._write_lock.release()

File: 5 - Coding Interview Workshops/Week12/Source Code/lesson.js
const { Worker, isMainThread, workerData } = require("worker_threads");
const Mutex = require("async-mutex").Mutex;

// Locks and Mutexes
if (isMainThread) {
  const sharedData = new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT);
  const lock = new Int32Array(sharedData);

  const worker = new Worker(__filename, { workerData: { sharedData } });
} else {
  const { sharedData } = workerData;
  const lock = new Int32Array(sharedData);

  Atomics.wait(lock, 0, 0);
  // Access shared resources
  Atomics.notify(lock, 0, 1);
}

// Semaphores
if (isMainThread) {
  const sharedData = new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT);
  const semaphore = new Int32Array(sharedData);

  const worker1 = new Worker(__filename, { workerData: { sharedData } });
  const worker2 = new Worker(__filename, { workerData: { sharedData } });
} else {
  const { sharedData } = workerData;
  const semaphore = new Int32Array(sharedData);

  Atomics.sub(semaphore, 0, 1);
  // Access shared resources
  Atomics.add(semaphore, 0, 1);
}

// Condition Variables
if (isMainThread) {
  const sharedData = new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT);
  const condition = new Int32Array(sharedData);

  const producerWorker = new Worker(__filename, {
    workerData: { sharedData, role: "producer" },
  });
  const consumerWorker = new Worker(__filename, {
    workerData: { sharedData, role: "consumer" },
  });
} else {
  const { sharedData, role } = workerData;
  const condition = new Int32Array(sharedData);

  if (role === "producer") {
    Atomics.store(condition, 0, 1);
    Atomics.notify(condition, 0, 1);
  } else if (role === "consumer") {
    Atomics.wait(condition, 0, 0);
    // Consume data
  }
}

// Scheduling Algorithms
function roundRobin(processes, quantum) {
  const readyQueue = [];
  processes.forEach((process) => readyQueue.push(process));

  while (readyQueue.length > 0) {
    const process = readyQueue.shift();
    // Execute process for the given quantum
    // If process is not finished, put it back in the queue
    if (!process.isFinished()) {
      readyQueue.push(process);
    }
  }
}

function priorityScheduling(processes) {
  const readyQueue = [];
  processes.forEach((process) => readyQueue.push(process));

  readyQueue.sort((a, b) => a.priority - b.priority);

  while (readyQueue.length > 0) {
    const process = readyQueue.shift();
    // Execute process until completion or preemption
  }
}

// Interview-Style Questions
let instance = null;
const lock = new Mutex();

class Singleton {
  constructor() {
    if (!instance) {
      lock.acquire();
      if (!instance) {
        instance = this;
      }
      lock.release();
    }
    return instance;
  }
}

class ConcurrentQueue {
  constructor() {
    this._queue = [];
    this._lock = new Mutex();
    this._condition = new Condition(this._lock);
  }

  put(item) {
    this._lock.acquire();
    this._queue.push(item);
    this._condition.notify();
    this._lock.release();
  }

  get() {
    this._lock.acquire();
    while (this._queue.length === 0) {
      this._condition.wait();
    }
    const item = this._queue.shift();
    this._lock.release();
    return item;
  }
}

class ReaderWriterLock {
  constructor() {
    this._readLock = new Mutex();
    this._writeLock = new Mutex();
    this._readers = 0;
  }

  acquireRead() {
    this._readLock.acquire();
    this._readers++;
    if (this._readers === 1) {
      this._writeLock.acquire();
    }
    this._readLock.release();
  }

  releaseRead() {
    this._readLock.acquire();
    this._readers--;
    if (this._readers === 0) {
      this._writeLock.release();
    }
    this._readLock.release();
  }

  acquireWrite() {
    this._writeLock.acquire();
  }

  releaseWrite() {
    this._writeLock.release();
  }
}
